### DISCUSSION TITLE
Color Scheme Map

### DISCUSSION THREAD

**Discussion by NullPlane** [SpartanJ/ecode#125] [Q&A] [answered]
May you create a Map for color schemes (UI-CSS & Syntax Highlighting),
There's a few entries I just couldn't find in the editor. 

**[BUG]** I did notice that for some reason `menu-font-active` does not work if preferred color scheme is dark (once bright and dark are implemented) or only one (no prefers) is implemented and there is no differentiation between bright and dark.

If you wish I may provide my schemes someday, but I'd have to do a lot of
refactoring and I'd have to publish them here.

**Comment by SpartanJ** [comment]
> May you create a Map for color schemes (UI-CSS & Syntax Highlighting),

You mean add a documentation for that? Sure.

> There's a few entries I just couldn't find in the editor.

Feel free to ask.

> [BUG] I did notice that for some reason menu-font-active does not work if preferred color scheme is dark (once bright and dark are implemented) or only one (no prefers) is implemented and there is no differentiation between bright and dark.

It works but it's not doing what you think. That's because the variable is only used to tint the icon colors and no the font color. But you can edit the CSS as you please though. For the light theme it was being used also to set the font color on mouse hover to a huge range of elements but I did just remove it a couple of days ago, but the [selector](https://github.com/SpartanJ/eepp/blob/develop/bin/assets/ui/breeze.css#L1108) is still valid as a general rule if you need it.

Usually base font color should be consistent over all elements. 

> If you wish I may provide my schemes someday, but I'd have to do a lot of refactoring and I'd have to publish them here.

That would be awesome! :)

**Comment by SpartanJ** [comment]
BTW the `breeze.css` is very easy to follow. If you need more information about the elements you can use the inspect tool that the editor provides: `Settings -> Window -> Inspect Elements`

**Comment by NullPlane** [comment, marked as answer]
**BUCKET** #661 and **THIS**

Mappings for CSS exist, hooray 🎉. However, nobody did that for colorschemes (highlighting) yet so, I thought, I'd make one.
This is everything as far as I've found it for the docs. I am not sure, it was quite a few things. Hope, I didn't mix things up. 
Just tell if something's missing.

---

## Colorschemes Map

The following keys are valid for the default set of syntax highlighting options in ecode.

| key                                   | element                                                                                                                           | <a href="#translatesTo">translation*</a>                                                                                                                                         | fallback/default (unspec) | syntax                                                                                                                                                                                            |
| ------------------------------------- | --------------------------------------------------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `background`                          | background color of editor area                                                                                                   | `--back`,<br/>`--tab-active`,<br/>`--term-back-color`                                                                                                                            | :x:                       | :x:                                                                                                                                                                                               |
| `widget_background`                   | background for panels and widgets that may spawn inside the editor                                                                | `--list-back`,<br/>`--list-row-active`,<br/>`--menu-back`,<br/>`--menu-font-active`                                                                                              | `gutter_background`       | :x:                                                                                                                                                                                               |
| `gutter_background`                   | background for gutter area (contains line numbers)                                                                                | :x:                                                                                                                                                                              | `background`              | :x:                                                                                                                                                                                               |
| `text`                                | text/code color for default visual text in the editor                                                                             | `--font`,<br/>`--term-font-color`,<br/>`--icon-line-hover`,<br/>`--menu-font`                                                                                                    | :x:                       | :x:                                                                                                                                                                                               |
| `caret`                               | caret color (blinking cursor)                                                                                                     | :x:                                                                                                                                                                              | :x:                       | :x:                                                                                                                                                                                               |
| `selection`                           | background color for selected text ranges                                                                                         | `--item-hover`,<br/>`--tab-hover`,<br/>`--icon-back-hover`                                                                                                                       | :x:                       | :x:                                                                                                                                                                                               |
| `line_highlight`<a id="lineHigh"></a> | background color of code line where caret is (current line) and selected lines (bg outside of selection box/text)                 | `--button-back`,<br/>`--separator`,<br/>`--slider-button`,<br/>`--scrollbar-border`,<br/>`--tab-back`                                                                            | :x:                       | :x:                                                                                                                                                                                               |
| `selection_region`                    | highlighted regions in batch operations such as find/replace                                                                      | `--scrollbar-hback-hover`                                                                                                                                                        | `selection`               | :x:                                                                                                                                                                                               |
| `matching_bracket`                    | background highlighting of matchable operators such as bracket pairs or quotes                                                    | :x:                                                                                                                                                                              | `selection`               | :x:                                                                                                                                                                                               |
| `matching_selection`                  | selecting text will highlight the same string sequence reoccurring in the document. Background color for reoc. text.              | :x:                                                                                                                                                                              | `selection`               | :x:                                                                                                                                                                                               |
| `matching_search`                     | background color for highlighted search/find results                                                                              | :x:                                                                                                                                                                              | :x:                       | :x:                                                                                                                                                                                               |
| `line_number`                         | color of line numbers in the gutter (inactive)                                                                                    | `--button-border`,<br/>`--slider-back`,<br/>`--tab-line`,<br/>`--menu-border`                                                                                                    | :x:                       | :x:                                                                                                                                                                                               |
| `line_number2`                        | color of the active line number(s) (see `line_highlight`)                                                                         | `--font-hint`,<br/>`--slider-border`,<br/>`--tab-close`,<br/>`--tab-font-inactive`,<br/>`--icon-line`,<br/>`--menu-font-disabled`,<br/>`--floating-icon`,<br/>`--disabled-color` | :x:                       | :x:                                                                                                                                                                                               |
| `whitespace`                          | color of visual whitespace markers (if enabled)                                                                                   | :x:                                                                                                                                                                              | `selection`               | :x:                                                                                                                                                                                               |
| `line_break_column`                   | color of the line breaking column (code length marker)                                                                            | `--disabled-border`                                                                                                                                                              | `selection`               | :x:                                                                                                                                                                                               |
| `suggestion`                          | color control for unselected suggestion items (autocomplete/lsp popup)                                                            | :x:                                                                                                                                                                              | :x:                       | `foreground,background`                                                                                                                                                                           |
| `suggestion_selected`                 | color control for selected suggestion items                                                                                       | :x:                                                                                                                                                                              | :x:                       | `foreground,background`                                                                                                                                                                           |
| `suggestion_scrollbar`                | scrollbar in suggestion box if list is large enough                                                                               | `--scrollbar-button`                                                                                                                                                             | `line_highlight`          | :x:                                                                                                                                                                                               |
| `minimap_background`                  | background of the right code minimap area                                                                                         | :x:                                                                                                                                                                              | `background`              | transparency suggested (e.g. `#282828AA`)                                                                                                                                                         |
| `minimap_visible_area`                | in the code minimap, there is a rectangle marking the currently visible section. background of said rectangle.                    | `droppable-hovering-color`                                                                                                                                                       | :x:                       | :x:                                                                                                                                                                                               |
| `minimap_current_line`                | the color of `line_highlight` on the minimap                                                                                      | :x:                                                                                                                                                                              | :x:                       | :x:                                                                                                                                                                                               |
| `minimap_hover`                       | the responsive color change when hovering mouse pointer over `minimap_visible_area`                                               | :x:                                                                                                                                                                              | :x:                       | :x:                                                                                                                                                                                               |
| `minimap_selection`                   | color of `selection` on the minimap                                                                                               | `--highlight-primary`                                                                                                                                                            | :x:                       | :x:                                                                                                                                                                                               |
| `minimap_highlight`                   | color of search/find highlights (`matching_search`) on the minimap                                                                | :x:                                                                                                                                                                              | :x:                       | :x:                                                                                                                                                                                               |
| `error`                               | color of message and wavy underline in error feedback (lsp, lint, ...)                                                            | `--tab-close-hover`,<br/>`--icon-back-alert`,<br/>`--theme-error`                                                                                                                | :x:                       | :x:                                                                                                                                                                                               |
| `warning`                             | color of message and wavy underline in warnings                                                                                   | `--theme-warning`                                                                                                                                                                | :x:                       | :x:                                                                                                                                                                                               |
| `notice`                              | color of message in informative notices                                                                                           | :x:                                                                                                                                                                              | :x:                       | :x:                                                                                                                                                                                               |
| `normal`                              | all canonical text/default code in code files (identifiers, variables, ...)                                                       | :x:                                                                                                                                                                              | :x:                       | :x:                                                                                                                                                                                               |
| `symbol`                              | operators, separators, symbols, punctuation in code                                                                               | `--icon`                                                                                                                                                                         | `normal`                  | :x:                                                                                                                                                                                               |
| `comment`                             | code comments; inline & multiline                                                                                                 | :x:                                                                                                                                                                              | :x:                       | :x:                                                                                                                                                                                               |
| `keyword`                             | keywords in comments (if, for, while, static, ...)                                                                                | :x:                                                                                                                                                                              | :x:                       | :x:                                                                                                                                                                                               |
| `keyword2` dep.;<br/>`type`           | types, declarations, primitives (char, int, float, ...)                                                                           | `--primary`,<br/>`--font-highlight`,<br/>`--tab-font-active`,<br/>`--icon-active`                                                                                                | :x:                       | :x:                                                                                                                                                                                               |
| `keyword3` dep.;<br/>`parameter`      | arguments/parameters in callable methods                                                                                          | :x:                                                                                                                                                                              | `symbol`                  | :x:                                                                                                                                                                                               |
| `number`                              | numeric value ($1$, $3.1415$, $0\mathrm x282828\mathrm{AA}$, ...)                                                                 | :x:                                                                                                                                                                              | :x:                       | :x:                                                                                                                                                                                               |
| `literal`                             | boolean literals and special constants resembling a valuable meaning ($\mathtt{True}$, $\mathtt{Null}$, $\mathtt{undefined}$,...) | :x:                                                                                                                                                                              | :x:                       | :x:                                                                                                                                                                                               |
| `string`                              | string literals (_"Goodbye, World!",_ ...)                                                                                        | :x:                                                                                                                                                                              | :x:                       | :x:                                                                                                                                                                                               |
| `operator`                            | mathematical, logical or language-agnostic operators (`+-*/=!=&&\|\|`)                                                            | :x:                                                                                                                                                                              | :x:                       | :x:                                                                                                                                                                                               |
| `function`                            | the name of a function (often before `()` parentheses)                                                                            | `--theme-success`                                                                                                                                                                | :x:                       | :x:                                                                                                                                                                                               |
| `link`                                | color of a callable/clickable hyperlink                                                                                           | :x:                                                                                                                                                                              | `function`                | :x:                                                                                                                                                                                               |
| `link_hover`                          | colors and theming when hyperlink is hovered by mouse pointer                                                                     | :x:                                                                                                                                                                              | `function`                | list color-code $\rightarrow$ transparency $\rightarrow$ shaded $\rightarrow$ underlined (modifiers enabled with adding the respective keyword); <br/>e.g. `#282828,transparent,shadow,underline` |
| `preprocessor`                        | color of preprocessor directives (`#include`, `{$R *.dfm}`, ...)                                                                  | :x:                                                                                                                                                                              | :x:                       | :x:                                                                                                                                                                                               |

Plugins may use custom keys.

## Operators/Valid values

Next to the colors, every highlighted object can be suited with an array of modifiers to change the appearance of the text or object. These are entered in a comma-separated list without spacing. 

**Available modifiers are:**

- `bold`
- `italic`
- `underline`
   - alt. `underlined`
- `strikethrough`
- `shadow` (outline)
- `regular` (explicitly unstyled)
- `outline(<color>, <thickness>)`
   - e.g. `outline(#cccccc, 0.7)`

> [!tip]
> 
> $\mathtt{color\\_code,modifier_1,\ldots,modifier_n}$
> 
> e.g. `link_hover = #4CE0BC,bold,underline,outline(#ffffff, 0.1)`

## Translations <a id="translatesTo"></a>

See #661; a syntax colorscheme may be [translated to the editor](https://github.com/SpartanJ/ecode/issues/661#issuecomment-3764904659).

**Reply by NullPlane** [reply]
As far as I see, this thread could almost be closed now, no? That's it then.

**Reply by SpartanJ** [reply]
AI got most of the things right but there are some hallucinated rules hehe. 

The GUI color scheme spec is already in the documentation, I'm doing several changes currently for the next version / release, updated version is [here](https://github.com/SpartanJ/ecode/blob/develop/docs/uicustomizations.md#gui-color-schemes-spec-map).
As for the mapping of the editor color scheme to GUI color scheme it's pretty straightforward within the [implementation](https://github.com/SpartanJ/eepp/blob/develop/src/tools/ecode/colorschemetranslator.cpp#L9), I haven't added any documentation yet because I'm still making changes.

I think the new feature at #661 adds a tone of flexibility to the editor styling, I'm quite happy with the result, I didn't expect to find a reasonable mapping for the colors but it's good enough.

Yes, this thread can be closed, I'll keep everything in the documentation.

**Reply by NullPlane** [reply]
Yes. But I manually rewrote the thing. Which keys are wrong?
The focus is not the CSS, I did that _"optionally"._ It's focused on
`colorschemes.conf`. Before deepwiki existed, I just had a python 
file scanning all keys from the existing conf because I was looking
for reference to create a syntax highlighting scheme. That does
not guarantee that all keys are used there and thus, listed. Also, not
guaranteed they're interpreted correctly by a designer without any 
description. Did I overlook that already existing? I see only CSS 
specs. Problem is, I sincerely wish for ecode to gain many more 
users but many won't look at any self a documenting code but
rather provided docs. I didn't have time but I still desperately needed
my own schemes and I am jot familiar with CPP and navigating
large projects I didn't write by myself so that was a swift solution,  
> _"[...] hehe"_ :D

gotta admit though, nowadays, it's much better. No more 
`keyword`, `keyword2`, `keyword3` :)

**Reply by SpartanJ** [reply]
> Which keys are wrong?

Just `preprocessor` does not exists, but I saw where did you got it, it was in one color scheme, probably while porting it i forgot to remove it. I ended up removing it yesterday.

The rest is great! I'll move this table into the documentation later. Thank you very much for documenting it!

**Reply by SpartanJ** [reply]
BTW if you have any cool color scheme, feel free to share it, I could add it with the default themes.

**Reply by NullPlane** [reply]
Can't wait!! I am a noob if it comes to UX design and I am striving
to shore something up since a long time. I want to make Grayskale a
project though like others do such that people can port it to their fav app.
However, Idk how to do specifications, how to get contrasts just right and
find the right shades for the right place and I fixate too much being a little
too perfectionist. If u know the route, I'd really appreciate every tip.

**Reply by NullPlane** [reply]
Oh! And alternatively, there is this scheme which I have previously "edited out" and made into the table... 
I don't know whether YOU prefer the table. I always like tabular data but the markdown tables are slightly 
crammed bc they r bound to a fixed width/maximums in GitHub so Idk whether you want to keep the 
translations in... Here is the unedited view just if you prefer it.

---

### Window/UI

- **`type`** → `--primary`: Primary accent color used throughout the UI (buttons on hover, selected items, active elements)
- **`background`** → `--back`: Main background color for windows, panels, and the application
- **`text`** → `--font`: Main text color for all UI elements
- **`text`** → `--term-font-color`: Terminal text color
- **`type`** → `--font-highlight`: Highlighted text color
- **`line_number2`** → `--font-hint`: Hint text color (placeholder text)

### Buttons

- **`line_highlight`** → `--button-back`: Button background color
- **`line_number`** → `--button-border`: Button border color

### Lists & spawnable panels

- **`widget_background`** → `--list-back`: Background for lists, dropdowns, text inputs, and panels
- **`widget_background`** → `--list-row-active`: Active/selected row text color
- **`line_highlight`** → `--separator`: Separator lines between UI elements
- **`selection`** → `--item-hover`: Hover state for list items and interactive elements

### Sliders

- **`line_number`** → `--slider-back`: Slider track background
- **`line_highlight`** → `--slider-button`: Slider thumb/button
- **`line_number2`** → `--slider-border`: Slider border

### Scrollbars

- **`line_highlight`** → `--scrollbar-border`: Scrollbar border
- **`suggestion_scrollbar`** → `--scrollbar-button`: Scrollbar thumb/button
- **`selection_region`** → `--scrollbar-hback-hover`: Scrollbar hover background

### Tabs

- **`line_highlight`** → `--tab-back`: Inactive tab background
- **`line_number`** → `--tab-line`: Tab border/separator lines
- **`background`** → `--tab-active`: Active tab background
- **`selection`** → `--tab-hover`: Tab hover background
- **`line_number2`** → `--tab-close`: Tab close button color
- **`error`** → `--tab-close-hover`: Tab close button hover color
- **`type`** → `--tab-font-active`: Active tab text color
- **`line_number2`** → `--tab-font-inactive`: Inactive tab text color

### Icons

- **`symbol`** → `--icon`: Default icon color
- **`type`** → `--icon-active`: Active/selected icon color
- **`selection`** → `--icon-back-hover`: Icon background on hover
- **`line_number2`** → `--icon-line`: Icon line color
- **`text`** → `--icon-line-hover`: Icon line color on hover
- **`error`** → `--icon-back-alert`: Alert icon background (e.g., window close)

### Menus & dialogs

- **`widget_background`** → `--menu-back`: Menu/dialog background
- **`text`** → `--menu-font`: Menu text color
- **`line_number`** → `--menu-border`: Menu border color
- **`widget_background`** → `--menu-font-active`: Active menu item text color
- **`line_number2`** → `--menu-font-disabled`: Disabled menu item text color

### Term

- **`background`** → `--term-back-color`: Terminal background
- **`line_number2`** → `--floating-icon`: Floating icon color

### Status & outputs

- **`minimap_selection`** → `--highlight-primary`: Primary highlight color
- **`line_number2`** → `--disabled-color`: Disabled element color
- **`line_break_column`** → `--disabled-border`: Disabled element border
- **`error`** → `--theme-error`: Error status color
- **`warning`** → `--theme-warning`: Warning status color
- **`function`** → `--theme-success`: Success status color
- **`minimap_visible_area`** → `droppable-hovering-color`: Drag & drop hover indicator

**Reply by NullPlane** [reply]
> The rest is great! I'll move this table into the documentation later. Thank you very much for documenting it!

I'll re-open the discussion until it is in there. Sorry, I am a bit quick to close stuff :)

**Reply by SpartanJ** [reply]
Thanks again! I've added it to the [documentation](https://github.com/SpartanJ/ecode/blob/main/docs/uicustomizations.md#editor-color-scheme-keys-map). It looks kinda weird, I'll improve it later but it's a start. Feel free to send PRs directly to improve it.
